### PR TITLE
Member expiration backend

### LINKED ISSUE #230: Member access expiration - backend
# Summary

Currently the group/project members functionality doesn't have an expiration field which denies access to an expired membership. Add in functionality so that the `members` table has an `expires_at` column and denies access to a member trying to access a resource when the membership has expired.

1) Add column to `members` table
2) Add scope non_expired to `member` model which will return members where the `expires_at is not null` and where `expires_at is > Time.zone.today`
3) Update calls in `member.rb` that are using `Member.for_namespace_and_ancestors` to determine access to include the `not_expired` scope
4) Add testing for model and controllers, to ensure only non expired members can access a group/project

**Note: The members -> groups tab already has similar functionality so the code for that page can be looked through for examples**

# Acceptance Criteria

Members which have an expires_at set should not be able to access a group or project if their membership has expired

### PR DESCRIPTION
**Pull request by ksierks**
## What does this PR do and why?
Fixes #230.
- Added a new `expires_at` column in the members table.
- Created a new scope `not_expired` for members.
- Updated policies to only retrieve non-expired members.

## Screenshots or screen recordings
N/A

## How to set up and validate locally
There are a couple ways to test:
1. Call the new scope via console
`Member.for_namespace_and_ancestors(@group).not_expired` or `Member.for_namespace_and_ancestors(@project.namespace).not_expired` and verify the members.
2. Even though this is not the fronted PR, you can log in as the user with an expired membership to make sure groups and projects are not displayed. Some areas to check are:
    - the groups/projects dashboards
    - transferring groups/projects drop-down list

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/models/member.rb
Comment (deepsidhu85) [inline app/models/member.rb:129]: We only want to return here if memberships exist from line 126. Otherwise this will throw an error when trying to do a `min`. 
To test this out you can create a new user, as `admin@email.com` add this user to `Bacillus`. Logout and login with this new user. Try to visit group `Yersinia`. You will see the error is caused by the `membership.last` in `membership.last.access_level`

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Just have a few comments to start. Also, the `group_policy` and `project_policy` scopes will need to be updated to not return groups/projects for which a user's membership has expired. This applies to groups and projects which may also be linked to one or more of the user's groups

File: app/models/member.rb
Comment (ksierks) [reply app/models/member.rb]: 5539d7984bd980ba188cf8fdee9e704a74113d8d

File: app/models/member.rb
Comment (deepsidhu85) [reply app/models/member.rb]: We still need to use the `.not_expired` when getting the memberships. The return where we get the minimum access level of the two needs to be wrapped in an if statement which checks if this membership array has a count

**Comment by deepsidhu85** [comment]
> Waiting for #319 to merge, so I can continue to update policies with the not_expired scope.

This PR doesn't need to be on hold right now. We can still update the `linked_groups` query in the `group_policy` -> `relation` scope. The `project_policy` `relation` scope also needs to be updated check if the user has a membership which isn't expired (`direct_projects`, `group_projects`, and `linked_projects` queries)

File: app/models/member.rb
Comment (ksierks) [reply app/models/member.rb]: f33e5a21dec476898a956cf201c0cbeb26e57744

File: test/policies/namespace_policy_test.rb
Comment (deepsidhu85) [inline test/policies/namespace_policy_test.rb:12]: This comment can be removed. That is correct that personal projects cannot be expired

File: test/policies/project_policy_test.rb
Comment (deepsidhu85) [inline test/policies/project_policy_test.rb:88]: This comment can also be removed

File: test/policies/project_policy_test.rb
Comment (deepsidhu85) [inline test/policies/project_policy_test.rb:158]: And this comment as well

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This is looking good. Just a few comments below. I still need to test this out

File: test/policies/namespace_policy_test.rb
Comment (ksierks) [reply test/policies/namespace_policy_test.rb]: 3a2aeb09e85eb371b5b577b45d56a3ecd01938fc

File: test/policies/project_policy_test.rb
Comment (ksierks) [reply test/policies/project_policy_test.rb]: 3a2aeb09e85eb371b5b577b45d56a3ecd01938fc

File: app/models/member.rb
Comment (deepsidhu85) [inline app/models/member.rb:123]: This return statement is not required here. It will end up giving the member this access level even though their membership has expired in which they are a member.

To test this out:
1) Create a new user
2) Login as admin@email.com
3) Create 2 groups, `Group A`, `Group B`
4) Add the new user to `Group B` with `MAINTAINER` access
5) Invite `Group B` to `Group A` with a group access level of `ANALYST` and do not set an expiration on the group link
6) Logout and login as the user created above
7) Verify you can see Group A and Group B, and that you can access them
8) In rails console expire the membership for the user in `Group B` with `m = Member.last` `m.expires_at = Date.today - 1` `m.save!`
9) You shouldn't see any groups listed on the groups dashboard, however if you visit `http://localhost:3000/group-a` you will still be able to access the `Group A` even though the membership for the user has expired in `Group B` which was linked to `Group A`

Remove this return fixes the issue and the user can no longer access `http://localhost:3000/group-a` as is expected

Please also add a test for this case to ensure that the user cannot access the group through a namespace group link when their membership has expired in their own group which was linked to this group

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
After testing it out I found a bug, I provided an example to test it

File: app/models/member.rb
Comment (ksierks) [reply app/models/member.rb]: Thank you, I was able to reproduce this bug locally. The expired groups were not listed in the dashboard table, but the user was still able to access them. Please see 3e1d2dc2566a3365559b01de88d62345afee898b for the fix and quick test.

**Review by ChrisHuynh333 — APPROVED** [review]
Works as described for me. Just confirming Deeps comment above.

File: app/models/member.rb
Comment (deepsidhu85) [reply app/models/member.rb]: Thanks!

**Review by deepsidhu85 — APPROVED** [review]
Looks good to me and working as described!

**Pull request merged**
